packages/pg-protocol/src/parser.ts

```diff
@@ -102,19 +113,94 @@ export class Parser {
+    })
+  }
 
-    if (offset === combinedBuffer.byteLength) {
+  private mergeBuffer(buffer: Buffer): CombinedBuffer {
+    let combinedBuffer = buffer
+    let combinedBufferLength = buffer.byteLength
+    let combinedBufferOffset = 0
+    let reuseRemainingBuffer = this.remainingBufferLength > 0
+    if (reuseRemainingBuffer) {
+      const newLength = this.remainingBufferLength + combinedBufferLength
+      const newFullLength = newLength + this.remainingBufferOffset
+      if (newFullLength > this.remainingBuffer.byteLength) {
+        // We can't concat the new buffer with the remaining one
+        let newBuffer: Buffer
+        if (newLength <= this.remainingBuffer.byteLength && this.remainingBufferOffset >= this.remainingBufferLength) {
+          // We can move the relevant part to the beginning of the buffer instead of allocating a new buffer
+          newBuffer = this.remainingBuffer
+        } else {
+          // Allocate a new larger buffer
+          let newBufferLength = this.remainingBuffer.byteLength * 2
+          while (newLength >= newBufferLength) {
+            newBufferLength *= 2
+          }
+          newBuffer = Buffer.allocUnsafe(newBufferLength)
+        }
+        // Move the remaining buffer to the new one
+        this.remainingBuffer.copy(
+          newBuffer,
+          0,
+          this.remainingBufferOffset,
+          this.remainingBufferOffset + this.remainingBufferLength
+        )
+        this.remainingBuffer = newBuffer
+        this.remainingBufferOffset = 0
+      }
+      // Concat the new buffer with the remaining one
+      buffer.copy(this.remainingBuffer, this.remainingBufferOffset + this.remainingBufferLength)
+      combinedBuffer = this.remainingBuffer
+      combinedBufferLength = this.remainingBufferLength = newLength
+      combinedBufferOffset = this.remainingBufferOffset
+    }
+    const combinedBufferFullLength = combinedBufferOffset + combinedBufferLength
+    return {
+      combinedBuffer,
+      combinedBufferOffset,
+      combinedBufferLength,
+      reuseRemainingBuffer,
+      combinedBufferFullLength,
+    }
+  }
+
+  private consumeBuffer({
+    combinedBufferOffset,
+    combinedBufferFullLength,
+    reuseRemainingBuffer,
+    combinedBuffer,
+    combinedBufferLength,
+  }: CombinedBuffer) {
+    if (combinedBufferOffset === combinedBufferFullLength) {
+      // No more use for the buffer
       this.remainingBuffer = emptyBuffer
+      this.remainingBufferLength = 0
+      this.remainingBufferOffset = 0
     } else {
-      this.remainingBuffer = combinedBuffer.slice(offset)
+      this.remainingBufferLength = combinedBufferFullLength - combinedBufferOffset
+      if (reuseRemainingBuffer) {
+        // Adjust the cursors of remainingBuffer
+        this.remainingBufferOffset = combinedBufferOffset
+      } else {
+        // To avoid side effects, copy the remaining part of the new buffer to remainingBuffer with extra space for next buffer
```

Comment:
@brianc this part is the only thing I think I can still improve. The question is if I can use the input buffer for the next run without anyone modifying it out of scope. What do you think?

Reply:
what do you mean by anyone modifying it out of scope? The only place in the whole codebase where we deal w/ raw input buffers from the socket is here in `parser.ts` and its not exposed to the user ever so any type of modifications you do here are safe in that no one else should (or can AFAIK) ever modify input buffers from the socket.

Reply:
It will be better here to use the existing input buffer for the next run instead of copying it over. What you say is that it is safe for me to do (as far as you know) which is great! I will change it then.
I was worried that who ever passes that buffer, can decide themselves to resuse that buffer and then it will mess up with my code...

Reply:
the buffer is passed in [here](https://github.com/brianc/node-postgres/blob/master/packages/pg/lib/connection.js#L104) as the whole `net.Socket`

The only place I could imagine this might cause issues is w/ some of the tests which may incorrectly be reusing buffers though nothing comes to mind.  Some of the tests are old and kinda wonky so you could just try & run the tests...as long as they pass we're definitely good. The raw buffers of the socket are never exposed anywhere outside of the parser.

Reply:
Great will do